benchmarks/cpp/src/qsort.cpp

```diff
@@ -0,0 +1,79 @@
+#include <algorithm>
+#include <array>
+#include <asio.hpp>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <vector>
+
+using namespace asio;
+using namespace std;
+using namespace std::chrono;
```

Comment:
Could the functions/namespaces being used be qualified instead of brought into scope like this? I'm having a bit of issue figuring out where `async` and `launch::async` come from and how `ctx.run()` is connected to them to wait for the tasks to complete.

Reply:
`std::launch::async` force the asynchronous operation while the asio loop event with steady_timer making the function awaitable on each call. ctx.run (is blocking - threadsafe).

Ref:
https://think-async.com/Asio/asio-1.24.0/doc/asio/reference/io_context.html#asio.reference.io_context.synchronous_and_asynchronous_operations

Reply:
From the link, it looks like you should have an io_context passed to all quickSorts which is posted() to for spawning subsequent quickSorts, with a context.run() at main

Reply:
Putting `ctx.run` in main would crash and probably still require release or reset `ctx.reset` before re-running on each call.

Reply:
Why? ctx.run() should invoke the  quicksorts, which should dispatch/queue moer, which should runs those, until all quicksorts completes.

Reply:
The problem at the moment from what I analyzed hangs during the call of the qsort function passing `ctx` (recursively).

So having the local context could be instantiated only after calling the reference lambda functions.

Reply:
I understand it this way:

**Local context:**

```
qsort(...):
   ctx
   lambda(qsort):
      ctx.run
   dispatch1(lambda) //ctx (run/stop)
   dispatch2(lambda) //ctx (run/stop)

    ctx.run/stop // will hold until it exits the operation.
```
**Main context:**

```
qsort(...,ctx):
   lambda(qsort,ctx):
      ctx.run
   dispatch1(lambda) //ctx (run)
   dispatch2(lambda) //ctx (run)

    ctx.run // recursive reference will crash the ctx (no stop, just run)
```

Reply:
Fixed: conceptual context.

Reply:
Under main-context, what does it mean that recursive context crashes the ctx?

Reply:
When executing after the qsort operation, the referenced ctx will simply not execute in the recursion.
Perhaps I would need to add a threadpool to instantiate the main context together with qsort.

If you execute the main context before to call qsort, the execution state will not be shared.

Reply:
I was thinking it runs like this:
```
main:
  qsort(0, 4):
    enqueue(qsort, 0, 2)
    enqueue(qsort, 2, 4)
  ctx.run():
    dequeue qsort(0, 2):
      enqueue(qsort, 0, 1)
      enqueue(qsprt, 1, 2)
    dequeue qsort(2, 4):
      enqueue(qsort, 2, 3)
      enqueue(qsort, 3, 4)
    dequeue(qsorts (0, 1), (1, 2), (2, 3), (3, 4))
    empty so stop
```

Reply:
Not work (deadlock)

Reply:
@kprotty , new commit w/ `ctx.run` on `main()`.